api/src/test/java/org/pmiops/workbench/audit/ActionAuditServiceTest.java

```diff
@@ -0,0 +1,116 @@
+  private ActionAuditEvent event2;
+  private ActionAuditService actionAuditService;
+
+  @Before
+  public void setUp() {
+    actionAuditService = new ActionAuditServiceImpl(mockLogging);
+    actionId = ActionAuditService.newActionId();
+
+    // ordinarily events sharing an action would have more things in common than this,
+    // but the schema doesn't require it
+    event1 =
+        new ActionAuditEvent.Builder()
+            .setAgentEmail("a@b.co")
+            .setTargetType(TargetType.DATASET)
+            .setTargetId(1L)
+            .setAgentType(AgentType.USER)
+            .setAgentId(AGENT_ID_1)
+            .setActionId(actionId)
+            .setTargetProperty("foot")
+            .setPreviousValue("bare")
+            .setNewValue("shod")
+            .setTimestamp(System.currentTimeMillis())
+            .build();
+
+    event2 =
+        new ActionAuditEvent.Builder()
+            .setAgentEmail("f@b.co")
+            .setTargetType(TargetType.DATASET)
+            .setTargetId(2L)
+            .setAgentType(AgentType.USER)
+            .setAgentId(AGENT_ID_2)
+            .setActionId(actionId)
+            .setTargetProperty("height")
+            .setPreviousValue("yay high")
+            .setNewValue("about that tall")
+            .setTimestamp(System.currentTimeMillis())
+            .build();
+  }
+
+  @Test
+  public void testSendsSingleEvent() {
+    actionAuditService.send(event1);
+    verify(mockLogging).write(logEntryListCaptor.capture());
+    List<LogEntry> entryList = logEntryListCaptor.getValue();
+    assertThat(entryList.size()).isEqualTo(1);
+    LogEntry entry = entryList.get(0);
+    assertThat(entry.getPayload().getType()).isEqualTo(Payload.Type.JSON);
+    JsonPayload jsonPayload = entry.getPayload();
+    assertThat(jsonPayload.getDataAsMap().size()).isEqualTo(11);
+
+    Map<String, Object> payloadMap = jsonPayload.getDataAsMap();
+    assertThat(payloadMap.get(AuditColumn.NEW_VALUE.name())).isEqualTo("shod");
+    // Logging passes numeric json fields as doubles when building a JsonPayload
+    assertThat(payloadMap.get(AuditColumn.AGENT_ID.name())).isEqualTo((double) AGENT_ID_1);
+  }
+
+  @Test
+  public void testSendsMultipleEvents() {
+    actionAuditService.send(ImmutableList.of(event1, event2));
+    verify(mockLogging).write(logEntryListCaptor.capture());
+    List<LogEntry> entryList = logEntryListCaptor.getValue();
+    assertThat(entryList.size()).isEqualTo(2);
+
+    final ImmutableList<JsonPayload> payloads =
+        entryList.stream()
+            .map(LogEntry::getPayload)
+            .filter(p -> ((Payload) p).getType() == Type.JSON)
+            .map(p -> (JsonPayload) p)
+            .collect(ImmutableList.toImmutableList());
+
+    assertThat(
+            payloads.stream()
+                .map(JsonPayload::getDataAsMap)
+                .map(entry -> entry.get(AuditColumn.ACTION_ID.name()))
+                .distinct()
+                .count())
+        .isEqualTo(1);
```

Comment:
is the goal of this test to ensure that duplicate events are not recorded? If so, I think that could be useful info to append to the test name

Reply:
Renamed to `testSendsMultipleEventsAsSingleAction`. We still have multiple events, but they're not considered duplicated. Just grouped with the actionId column.